ipyparallel/cluster/launcher.py

```diff
@@ -1363,15 +1357,16 @@ def _start_waiting(self):
 
     def wait_one(self, timeout):
         python_code = f"from ipyparallel.cluster.launcher import ssh_waitpid; ssh_waitpid({self.pid}, timeout={timeout})"
-        full_cmd = (
-            self.ssh_cmd
-            + self.ssh_args
-            # double-quote for ssh
-            + [self.location, "--", self.remote_python, "-c", f"'{python_code}'"]
-        )
-        out = check_output(full_cmd, input=None, start_new_session=True).decode(
-            "utf8", "replace"
-        )
+        # full_cmd = (
+        #     self.ssh_cmd
+        #     + self.ssh_args
+        #     # double-quote for ssh
+        #     + [self.location, "--", self.remote_python, "-c", f"'{python_code}'"]
+        # )
+        # out = check_output(full_cmd, input=None, start_new_session=True).decode(
+        #     "utf8", "replace"
+        # )
+        out = self._secure_ssh_sender().check_output_python_code(python_code)
```

Comment:
What's the difference between `self.ssh_sender` and `self._secure_ssh_sender`?

Reply:
`_secure_ssh_sender ` creates the `ssh_sender `object if needed. Maybe it should be named  differently, e.g. `construct_ssh_sender_if_needed`
Since I do not (fully) understand the object construction concept in _ipp_ it seemed to me the easiest way of creating the ssh-sender object (once). I'm pretty sure that I needed the object creation at two different locations:
`SSHLauncher.start` and  `SSHLauncher.wait_one` which is why I put it into a function.

Any suggestions to make this code part more elegant or better understandable?

Reply:
ok and thanks. But if I use `if self._ssh_sender:` the variable `self._ssh_sender` must be initialised somewhere. E.g. in the non-existing `__init__` function
```
    def __init__(self, **kwargs):
        super().__init__(kwargs)
        self._ssh_sender = None
```
or as _parameter_ (like e.g. `to_send`). Otherwise I would still need to use `if getattr(self, '_ssh_sender', None):` syntax. 
Which method do you prefer?

Reply:
Usually I'd initialize it to None at the class level right before the property declaration:

```python
_ssh_sender = None

@property
def ssh_sender(self):
    ...
```

 Or use `getattr(self, '_ssh_sender', None)` for the initial check so it doesn't need to be declared first.

Reply:
Changed as suggested (see 0601b51fa12834de2a14529d5119167bd59039e4)